### PR TITLE
[FLINK-13367] Add support for writeReplace in nested ClosureCleaner.

### PR DESCRIPTION
**Pull request by dmvk**
## What is the purpose of the change

Support for writeReplace in ClosureCleaner for Apache Beam [Avro Coder](https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/coders/AvroCoder.java#L215).

## Brief change log

  - *Added failing test case*
  - *Make usesCustomSerialiazation method search for writeReplace method.*

## Verifying this change

This change added tests and can be verified as follows:

  - *Added failing test case that uses writeReplace method for serializing payload*
## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (yes)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-java/src/test/java/org/apache/flink/api/java/functions/ClosureCleanerTest.java [hunk 1]
```diff
@@ -210,6 +210,13 @@ public void testRecursiveClass() {
 
 		ClosureCleaner.ensureSerializable(recursiveClass);
 	}
+
+	@Test
+	public void testWriteReplace() {
```
[the remark this answers is not in the document]
Comment (mxm) [reply flink-java/src/test/java/org/apache/flink/api/java/functions/ClosureCleanerTest.java]: I suppose `readResolve` requires `writeReplace`, so not necessary.

File: flink-java/src/test/java/org/apache/flink/api/java/functions/ClosureCleanerTest.java [hunk 2]
```diff
@@ -421,3 +428,59 @@ public Integer map(Integer value) throws Exception {
...
+		private Object writeReplace() {
+			return new SerializedPayload(payload.getRaw());
+		}
+
+		Payload get() {
+			return payload;
+		}
+	}
+
+	private static class SerializedPayload implements Serializable {
```
[the remark this answers is not in the document]
Comment (dmvk) [reply flink-java/src/test/java/org/apache/flink/api/java/functions/ClosureCleanerTest.java]: added test case for the non-recursive one

File: flink-java/src/test/java/org/apache/flink/api/java/functions/ClosureCleanerTest.java [hunk 3]
```diff
@@ -421,3 +428,59 @@ public Integer map(Integer value) throws Exception {
...
+		}
+
+		private Object readResolve() throws IOException, ClassNotFoundException {
+			return new Payload(raw);
+		}
+	}
+
+	private final SerializablePayload serializablePayload;
+
+	WithWriteReplace(Payload payload) {
```
[the remark this answers is not in the document]
Comment (dmvk) [reply flink-java/src/test/java/org/apache/flink/api/java/functions/ClosureCleanerTest.java]: Changed.

Anyway this seems to differ a lot throughout the code base. The way I'm trying to think about this is that static code should always appear before non-static. It is usually confusing if static and instance classes are mixed together.

- static properties
- static methods
- static classes
- instance properties
- instance methods
- instance classes

**Comment by dmvk** [comment]
@mxm Thanks for the review Max! I've incorporated your comments, can you please take a look again?

D.

File: flink-java/src/test/java/org/apache/flink/api/java/functions/ClosureCleanerTest.java [hunk 4]
```diff
@@ -234,10 +249,11 @@ public void writeObject(ObjectOutputStream o) {
...
 
-	static MapFunction<Integer, Integer> map1;
-	transient MapFunction<Integer, Integer> map2;
-	CustomMap map3;
-	LocalMap map4;
+	private static MapFunction<Integer, Integer> map1;
+
+	private transient MapFunction<Integer, Integer> map2;
+	private CustomMap map3;
+	private LocalMap map4;
```
Comment (mxm) [inline flink-java/src/test/java/org/apache/flink/api/java/functions/ClosureCleanerTest.java]: unrelated changes, but seems fair

**Comment by aljoscha** [comment]
Also looks good from my side! 👌 Only the commit title "Add support for writeReplace in nested ClosureCleaner." threw me off a bit. Maybe "Recognize writeReplace in ClosureCleaner" would be better.

**Comment by mxm** [comment]
Cool. Tests are green. @dmvk I'll adjust the commit description before merging this.

**Pull request closed without merging**